### PR TITLE
debug: add run statusbar item

### PR DESCRIPTION
**Pull request by vince-fugnitto**
#### What it does

Fixes: #5156

The following pull-request implements the _select and run_ debug statusbar item:
- includes a new preference `debug.showInStatusBar`
- the statusbar item triggers the prefix select and run command
- the statusbar item reflects the current debug configuration, and updates when the config changes, and when the preference value is updated

#### How to test

Verify that the debug status bar item is updated accordingly based on the preference value, and the configuration updates:
1. start the application, and open the debug view
2. preferences:
  - `never`: the debug status bar item is never displayed, if previously displayed it will be removed
  - `always`: the debug status bar item is always displayed (on app start as well)
  - `onFirstSessionStart`: the debug status bar item is displayed on a debug session is first started
3. verify that the debug status bar item is updated when a configuration is updated (label), and when the preference value is updated

---

_Example: updating the configuration changes the label_

<div align='center'>

debug-sb

</div>

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: vince-fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

File: packages/core/src/browser/status-bar/status-bar.tsx [hunk 1]
```diff
@@ -66,6 +66,7 @@ export interface StatusBar {
     setColor(color?: string): Promise<void>;
     setElement(id: string, entry: StatusBarEntry): Promise<void>;
     removeElement(id: string): Promise<void>;
+    exists(id: string): Promise<boolean>;
```
Comment (akosyakov) [inline packages/core/src/browser/status-bar/status-bar.tsx:69]: Do we really need it? You can just always update no as in other places?

File: packages/debug/src/browser/debug-prefix-configuration.ts [hunk 2]
```diff
@@ -49,15 +50,30 @@ export class DebugPrefixConfiguration implements CommandContribution, CommandHan
...
         id: 'select.debug.configuration',
         category: 'Debug',
         label: 'Select and Start Debugging'
     };
 
+    @postConstruct()
+    init(): void {
+        this.debugSessionManager.onDidStartDebugSession(() => this.updateStatusBar());
+        this.debugConfigurationManager.onDidChange(async () => {
+            const exists = await this.statusBar.exists(this.statusBarId);
```
Comment (akosyakov) [inline packages/debug/src/browser/debug-prefix-configuration.ts:70]: why is it necessary?

File: packages/debug/src/browser/debug-prefix-configuration.ts [hunk 3]
```diff
@@ -49,15 +50,30 @@ export class DebugPrefixConfiguration implements CommandContribution, CommandHan
...
+    readonly statusBarId = 'select-run-debug-statusbar-item';
 
     private readonly command: Command = {
         id: 'select.debug.configuration',
         category: 'Debug',
         label: 'Select and Start Debugging'
     };
 
+    @postConstruct()
+    init(): void {
```
Comment (akosyakov) [inline packages/debug/src/browser/debug-prefix-configuration.ts:67]: protected

File: packages/debug/src/browser/debug-prefix-configuration.ts [hunk 4]
```diff
@@ -119,4 +135,16 @@ export class DebugPrefixConfiguration implements CommandContribution, CommandHan
         this.debugConfigurationManager.current = { ...configuration };
         this.commandRegistry.executeCommand(DebugCommands.START.id);
     }
+
+    protected updateStatusBar(): void {
+        const text: string = this.debugConfigurationManager.current
+            ? this.debugConfigurationManager.current.configuration.name
+            : 'Start Debugging';
```
Comment (akosyakov) [inline packages/debug/src/browser/debug-prefix-configuration.ts:142]: in VS Code there is empty text if there are no configurations, try to delete launch.json

**Comment by akosyakov** [comment]
Is it configurable in VS Code?

**Comment by vince-fugnitto** [comment]
> Is it configurable in VS Code?

I was previously unaware of any configuration, but verifying I found the following preference which controls the behavior:

debug-statusbar

- [VS Code Implementation](https://github.com/microsoft/vscode/blob/b988f09b1e7c9aca37b8f312ba34c6422757617b/src/vs/workbench/contrib/debug/browser/debug.contribution.ts#L218-L223)

File: packages/debug/src/browser/debug-prefix-configuration.ts (hunk 3 above)
Comment (vince-fugnitto) [reply packages/debug/src/browser/debug-prefix-configuration.ts:71]: Updating the method to protected causes issues with the compatibility of `QuickOpenHandler` which is unprotected.

File: packages/debug/src/browser/debug-prefix-configuration.ts [hunk 5]
```diff
@@ -119,4 +143,39 @@ export class DebugPrefixConfiguration implements CommandContribution, CommandHan
...
+    /**
+     * Update the debug status bar element based on the current configuration.
+     */
+    protected updateStatusBar(): void {
+        const text: string = this.debugConfigurationManager.current
+            ? this.debugConfigurationManager.current.configuration.name
+            : '';
+        this.statusBar.setElement(this.statusBarId, {
+            alignment: StatusBarAlignment.LEFT,
+            text: '$(play) ' + text,
```
Comment (akosyakov) [inline packages/debug/src/browser/debug-prefix-configuration.ts:169]: space should be added only if text is not empty

File: packages/debug/src/browser/debug-prefix-configuration.ts [hunk 6]
```diff
@@ -49,15 +54,34 @@ export class DebugPrefixConfiguration implements CommandContribution, CommandHan
...
+    readonly statusBarId = 'select-run-debug-statusbar-item';
 
     private readonly command: Command = {
         id: 'select.debug.configuration',
         category: 'Debug',
         label: 'Select and Start Debugging'
     };
 
+    @postConstruct()
+    protected update(): void {
```
Comment (akosyakov) [inline packages/debug/src/browser/debug-prefix-configuration.ts:71]: `init` is better name, since it happens only once on the construction

Comment (vince-fugnitto) [reply packages/debug/src/browser/debug-prefix-configuration.ts:71]: @akosyakov unfortunately I cannot use `protected init` as it is present in the `QuickOpenHandler` for which we implement:

https://github.com/eclipse-theia/theia/blob/5818d507201f84701f99788c76a465028bca59b0/packages/core/src/browser/quick-open/prefix-quick-open-service.ts#L50-L55

Comment (akosyakov) [reply packages/debug/src/browser/debug-prefix-configuration.ts:71]: `initialize`? :D

Comment (vince-fugnitto) [reply packages/debug/src/browser/debug-prefix-configuration.ts:71]: > `initialize`? :D

haha sure :)

**Comment by akosyakov** [comment]
@vince-fugnitto could you assign someone to test it please?

**Comment by marcdumais-work** [comment]
> In this case, by reading the code I can see it (default: 'onFirstSessionStart')

You can't always rely on the default preference value in the code, since the default can easily be overridden in the application's `package.json`. 

> Could be in a separate PR: It would be nice to the end-user to know which value is the default value for any preferences when selecting it.

Sounds like you're suggesting a potential Preferences widget improvement? However pretty orthogonal to this PR here, so if you care, you should open a separate issue about it, if none exist yet.

**Comment by lmcbout** [comment]
> You can't always rely on the default preference value in the code, since the default can easily be overridden in the application's package.json.

 As a developper of Theia, I can open the file and see the default value. As a end-user, the only thing I see is in the preference setting while running Theia, I don't look inside the code to find the default value. I did an extra test and in VSCode, the default value is identified in the preference setting. If we try to align Theia with VSCode, then the default value is not identified 

I don't know about the other preferences default setting, i was just looking at the preference setting for this issue.

**Pull request merged**